File: Solutions/Maximum_Product_Subarray/Maximum_Product_Subarray.go
//The idea is, declaring two variables for tracking the minimum and the maximum product value. We need to keep track of minimum value for negative number. If the any ith (nums[i])number is a negative number then we need to swap between mimimun value and maximum value as in this case minimum value will be multiplied by a negative number and it will become a positive maximum value.
//For Deciding the min and max value in each iteration we have to look for the best option available means for max value we have to check maxvalue*nums[i] or nums[i] which is the max - which is the better option to choose similarly for the min value
//So basically it's dynamic programing problem where we have a take decision based upon previous reult and current option

func maxProduct(nums []int) int {
    minval := nums[0]
    maxval := nums[0]
    maxproduct := nums[0]
    for i := 1; i<len(nums); i++{
        if nums[i] < 0{
            minval, maxval = maxval, minval
        }
        minval = int(math.Min(float64(nums[i]),float64(minval*nums[i])))
        maxval = int(math.Max(float64(nums[i]),float64(maxval*nums[i])))
        if maxval > maxproduct{
            maxproduct = maxval
        }
    }
    return maxproduct
}

File: Solutions/Maximum_Product_Subarray/Maximum_Product_Subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    var n = nums.length;
    if (!n) return 0;
    var curMax = nums[0];
    var curMin = nums[0];
    var max = nums[0];
    for (var i = 1; i < n; i++) {
        var num = nums[i];
        var minProd = curMin * num;
        var maxProd = curMax * num;
        curMax = Math.max(num, minProd, maxProd);
        curMin = Math.min(num, minProd, maxProd);
        max = Math.max(max, curMax);
    }
    return max;
};

//Fastest
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    var max = nums[0];
    var min = nums[0];
    var result = max;
    for (var i = 1; i < nums.length; i += 1) {
        var current = nums[i];
        var preMax = max
        var max = Math.max(current, Math.max(preMax * current, min * current));
        var min = Math.min(current, Math.min(preMax * current, min * current));
        result = Math.max(max, result);
    }
    return result;
};